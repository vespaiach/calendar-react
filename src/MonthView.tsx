import './month.css';

import React, { useMemo } from 'react';

import { Event as EventType, MonthNum } from './type';
import { cx, dateToString, dateToTime, DAYS as DEFAULT_DAYS, MONTHS as DEFAULT_MONTHS } from './utils';

interface MonthViewProps {
    month: MonthNum;
    year: number;
    onClick?: (date: Date, events: EventType[], evt: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void;
    DateComponent?: typeof DateButton;
    style?: React.CSSProperties;
    className?: string;
    DAYS?: string[];
    MONTHS?: string[];
    events?: EventType[];
    colors?: {
        gridLine?: string;
        sheetBackground?: string;
        dayText?: string;
        grayedOutText?: string;
        normalText?: string;
        lightText?: string;
        todayCircle?: string;
        eventListDot?: string;
    };
}

export const MonthView = React.forwardRef<HTMLDivElement, MonthViewProps>(function MonthComponent(
    {
        month,
        year,
        style,
        className,
        onClick,
        DAYS = DEFAULT_DAYS,
        MONTHS = DEFAULT_MONTHS,
        DateComponent = DateButton,
        events,
        colors,
    },
    ref,
) {
    const today = new Date();
    const currentDateString = dateToString(today);

    const eventsByStartDate = useMemo(
        () =>
            (events || [])
                .sort((a, b) => a.start.getTime() - b.start.getTime())
                .reduce((acc, evt) => {
                    acc[dateToString(evt.start)] = acc[dateToString(evt.start)] || [];
                    acc[dateToString(evt.start)].push(evt);
                    return acc;
                }, {} as Record<string, EventType[]>),
        [events],
    );

    const dayEls = useMemo(() => {
        const start = new Date(year, month, 1);

        const nextMonth = new Date(start);
        nextMonth.setMonth(nextMonth.getMonth() + 1);

        if (start.getDay() > 0) {
            start.setDate(start.getDate() - start.getDay());
        }

        const curr = new Date();
        const els = [];
        let count = 0;
        while (start < nextMonth || count % 7 > 0) {
            count++;
            const dateString = dateToString(start);
            const isToday = dateString === currentDateString;
            const dateEvents = eventsByStartDate[dateString];
            const copyStartDate = new Date(start);

            els.push(
                <DateComponent
                    date={new Date(start)}
                    standout={isToday}
                    className={cx(
                        'flex flex-col p-4 border-t border-line',
                        count % 7 !== 0 && 'border-r',
                        start.getMonth() !== month && 'text-grayed-out',
                    )}
                    tabIndex={start.getMonth() !== month ? -1 : undefined}
                    key={start.getTime()}
                    onClick={(evt) => void onClick?.(copyStartDate, dateEvents, evt)}>
                    {start.getDate() === 1 ? (
                        <div className="text-left">{MONTHS[start.getMonth()]}</div>
                    ) : null}
                    {dateEvents?.map((evt, i) => (
                        <p
                            key={i}
                            className={cx(
                                'w-full relative truncate text-xs text-grayed-out pl-4 before:content-["*"]  before:text-xl before:absolute before:-top-1 before:left-0',
                                i === 0 ? 'mt-3' : 'mt-1',
                                start > curr && 'before:text-event-dot',
                            )}>
                            <strong>{`${dateToTime(evt.start)}`}</strong>
                            {` ${evt.title}`}
                        </p>
                    ))}
                </DateComponent>,
            );

            start.setDate(start.getDate() + 1);
        }

        return els;
    }, [month, year, onClick, currentDateString, eventsByStartDate, DateComponent]);

    const computedStyle = useMemo(() => {
        if (!colors || !Object.keys(colors).length) return style;

        const st: Record<string, unknown> = {};

        if (colors.dayText) {
            st['--mv-color-day-text'] = colors.dayText;
        }
        if (colors.eventListDot) {
            st['--mv-color-event-dot'] = colors.eventListDot;
        }
        if (colors.grayedOutText) {
            st['--mv-color-grayed-out-text'] = colors.grayedOutText;
        }
        if (colors.gridLine) {
            st['--mv-color-grid-line'] = colors.gridLine;
        }
        if (colors.lightText) {
            st['--mv-color-light-text'] = colors.lightText;
        }
        if (colors.normalText) {
            st['--mv-color-normal-text'] = colors.normalText;
        }
        if (colors.sheetBackground) {
            st['--mv-color-sheet-background'] = colors.sheetBackground;
        }
        if (colors.todayCircle) {
            st['--mv-color-today-circle'] = colors.todayCircle;
        }

        return { ...st, ...(style || {}) } as React.CSSProperties;
    }, [colors, style]);

    return (
        <div
            style={computedStyle}
            className={cx(
                'bg-sheet grid grid-rows-[1fr_repeat(5,_2fr)] grid-cols-7 auto-rows-[2fr] h-full text-normal',
                className,
            )}
            ref={ref}>
            {DAYS.map((name, i) => (
                <div
                    className={cx(i < 6 && 'border-r border-r-line', 'text-sm text-day text-center pt-2')}
                    key={name}>
                    {name}
                </div>
            ))}
            {dayEls}
        </div>
    );
});

interface DateComponentProps {
    date: Date;
    className?: string;
    tabIndex?: number;
    standout?: boolean;
    onClick?: React.MouseEventHandler<HTMLButtonElement>;
    children?: React.ReactNode;
}

export const DateButton = React.forwardRef<HTMLButtonElement, DateComponentProps>(function DateComponent(
    { date, className, tabIndex, standout, children, onClick },
    ref,
) {
    return (
        <button
            className={className}
            tabIndex={tabIndex}
            title={dateToString(date)}
            onClick={onClick}
            ref={ref}>
            {standout ? (
                <div className="w-9 h-9 flex justify-center items-center rounded-full bg-today text-light -mt-2 -ml-2">
                    <span>{date.getDate()}</span>
                </div>
            ) : (
                <span className="text-left">{date.getDate()}</span>
            )}
            {children}
        </button>
    );
});
