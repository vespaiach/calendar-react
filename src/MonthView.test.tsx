import React from 'react';
import { beforeEach, afterEach, test, expect, vi } from 'vitest';
import { act as actRender, create, ReactTestRenderer, ReactTestRendererJSON } from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import { render, unmountComponentAtNode } from 'react-dom';
import { Event as EventType } from './type';
import { DateButton, MonthView } from './MonthView';

let container: HTMLDivElement | null = null;
beforeEach(async () => {
    container = document.createElement('div');
    document.body.appendChild(container);
});

afterEach(() => {
    if (container) {
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    }
});

let component: ReactTestRenderer | null = null;
actRender(() => {
    component = create(<MonthView month={0} year={2022} />);
});

test('render DateButton components', () => {
    const dateButtonsInJanuary = component?.root.findAllByType(DateButton);
    expect(dateButtonsInJanuary).toHaveLength(42);
    const grayedOutdateButtonsInJanuary = component?.root
        .findAllByProps({ tabIndex: -1 })
        .filter((b) => b.type === 'button');
    expect(grayedOutdateButtonsInJanuary).toHaveLength(11);

    actRender(() => {
        component?.update(<MonthView month={1} year={2022} />);
    });
    const dateButtonsInFebuary = component?.root.findAllByType(DateButton);
    expect(dateButtonsInFebuary).toHaveLength(35);
    const grayOutdateButtonsInFebuary = component?.root
        .findAllByType('button')
        .filter((b) => b.props['tabIndex'] === -1);
    expect(grayOutdateButtonsInFebuary).toHaveLength(7);
});

test('mount to DOM', () => {
    act(() => {
        render(<MonthView month={1} year={2022} className="abc" />, container);
    });

    const root = container?.querySelector('>div');
    expect(root?.className).toBe(
        'bg-sheet grid grid-rows-[1fr_repeat(5,_2fr)] grid-cols-7 auto-rows-[2fr] h-full text-normal abc',
    );
    expect(root?.querySelectorAll('div[class~=text-day]')).toHaveLength(7);
});

test('click event', () => {
    const dt = new Date(2022, 0, 1, 12, 0, 0, 0);
    const nextDt = new Date(2022, 0, 2, 12, 0, 0, 0);
    const events = [
        { start: new Date(dt), minutes: 30, title: 'event title 1' },
        {
            start: new Date(nextDt.setHours(nextDt.getHours() + 1)),
            minutes: 30,
            title: 'event title 2a',
        },
        {
            start: new Date(nextDt.setHours(nextDt.getHours() + 1)),
            minutes: 30,
            title: 'event title 2b',
        },
    ];
    const clickEvent = vi.fn(() => 0);
    act(() => {
        render(<MonthView month={0} year={2022} events={events} onClick={clickEvent} />, container);
    });

    const btn = container?.querySelector('button[title="2022-01-01"]');
    expect(btn?.textContent).toBe('1January12:00 am event title 1');

    act(() => {
        btn?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });
    expect(clickEvent).toBeCalledTimes(1);
    expect(clickEvent.mock.calls[0]).toHaveLength(3);

    const returnedDate = clickEvent.mock.calls[0][0] as unknown as Date;
    expect(returnedDate.getFullYear()).toBe(2022);
    expect(returnedDate.getMonth()).toBe(0);
    expect(returnedDate.getDate()).toBe(1);

    const returnedEvents = clickEvent.mock.calls[0][1] as unknown as EventType[];
    expect(returnedEvents).toHaveLength(1);
    expect(returnedEvents[0].start.getFullYear()).toBe(2022);
    expect(returnedEvents[0].start.getMonth()).toBe(0);
    expect(returnedEvents[0].start.getDate()).toBe(1);
    expect(returnedEvents[0].start.getHours()).toBe(12);
    expect(returnedEvents[0].start.getMinutes()).toBe(0);
    expect(returnedEvents[0].minutes).toBe(30);
    expect(returnedEvents[0].title).toBe('event title 1');

    const btn2 = container?.querySelector('button[title="2022-01-02"]');
    expect(btn2?.textContent).toBe('21:00 pm event title 2a2:00 pm event title 2b');
    act(() => {
        btn2?.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(clickEvent).toBeCalledTimes(2);
    expect(clickEvent.mock.calls).toHaveLength(2);
    expect(clickEvent.mock.calls[1][1]).toHaveLength(2);
});

test('MonthView component snapshot', () => {
    const component = create(<MonthView month={0} year={2022} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
});

function toJson(component: ReactTestRenderer) {
    const result = component.toJSON();
    expect(result).toBeDefined();
    return result as ReactTestRendererJSON;
}
