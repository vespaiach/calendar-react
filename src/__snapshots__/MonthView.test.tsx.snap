// Vitest Snapshot v1

exports[`MonthView component snapshot 1`] = `
<div
  className="bg-sheet grid grid-rows-[1fr_repeat(5,_2fr)] grid-cols-7 auto-rows-[2fr] h-full text-normal"
>
  <div
    className="border-r border-r-line text-sm text-day text-center pt-2"
  >
    Sunday
  </div>
  <div
    className="border-r border-r-line text-sm text-day text-center pt-2"
  >
    Monday
  </div>
  <div
    className="border-r border-r-line text-sm text-day text-center pt-2"
  >
    Tuesday
  </div>
  <div
    className="border-r border-r-line text-sm text-day text-center pt-2"
  >
    Wednesday
  </div>
  <div
    className="border-r border-r-line text-sm text-day text-center pt-2"
  >
    Thursday
  </div>
  <div
    className="border-r border-r-line text-sm text-day text-center pt-2"
  >
    Friday
  </div>
  <div
    className="text-sm text-day text-center pt-2"
  >
    Saturday
  </div>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2021-12-26"
  >
    <span
      className="text-left"
    >
      26
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2021-12-27"
  >
    <span
      className="text-left"
    >
      27
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2021-12-28"
  >
    <span
      className="text-left"
    >
      28
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2021-12-29"
  >
    <span
      className="text-left"
    >
      29
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2021-12-30"
  >
    <span
      className="text-left"
    >
      30
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2021-12-31"
  >
    <span
      className="text-left"
    >
      31
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line"
    onClick={[Function]}
    title="2022-01-01"
  >
    <span
      className="text-left"
    >
      1
    </span>
    <div
      className="text-left text-sm"
    >
      January
    </div>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-02"
  >
    <span
      className="text-left"
    >
      2
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-03"
  >
    <span
      className="text-left"
    >
      3
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-04"
  >
    <span
      className="text-left"
    >
      4
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-05"
  >
    <span
      className="text-left"
    >
      5
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-06"
  >
    <span
      className="text-left"
    >
      6
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-07"
  >
    <span
      className="text-left"
    >
      7
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line"
    onClick={[Function]}
    title="2022-01-08"
  >
    <span
      className="text-left"
    >
      8
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-09"
  >
    <span
      className="text-left"
    >
      9
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-10"
  >
    <span
      className="text-left"
    >
      10
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-11"
  >
    <span
      className="text-left"
    >
      11
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-12"
  >
    <span
      className="text-left"
    >
      12
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-13"
  >
    <span
      className="text-left"
    >
      13
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-14"
  >
    <span
      className="text-left"
    >
      14
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line"
    onClick={[Function]}
    title="2022-01-15"
  >
    <span
      className="text-left"
    >
      15
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-16"
  >
    <span
      className="text-left"
    >
      16
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-17"
  >
    <span
      className="text-left"
    >
      17
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-18"
  >
    <span
      className="text-left"
    >
      18
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-19"
  >
    <span
      className="text-left"
    >
      19
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-20"
  >
    <span
      className="text-left"
    >
      20
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-21"
  >
    <span
      className="text-left"
    >
      21
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line"
    onClick={[Function]}
    title="2022-01-22"
  >
    <span
      className="text-left"
    >
      22
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-23"
  >
    <span
      className="text-left"
    >
      23
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-24"
  >
    <span
      className="text-left"
    >
      24
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-25"
  >
    <span
      className="text-left"
    >
      25
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-26"
  >
    <span
      className="text-left"
    >
      26
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-27"
  >
    <span
      className="text-left"
    >
      27
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-28"
  >
    <span
      className="text-left"
    >
      28
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line"
    onClick={[Function]}
    title="2022-01-29"
  >
    <span
      className="text-left"
    >
      29
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-30"
  >
    <span
      className="text-left"
    >
      30
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r"
    onClick={[Function]}
    title="2022-01-31"
  >
    <span
      className="text-left"
    >
      31
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2022-02-01"
  >
    <span
      className="text-left"
    >
      1
    </span>
    <div
      className="text-left text-sm"
    >
      February
    </div>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2022-02-02"
  >
    <span
      className="text-left"
    >
      2
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2022-02-03"
  >
    <span
      className="text-left"
    >
      3
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line border-r text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2022-02-04"
  >
    <span
      className="text-left"
    >
      4
    </span>
  </button>
  <button
    className="flex flex-col p-4 border-t border-line text-grayed-out"
    onClick={[Function]}
    tabIndex={-1}
    title="2022-02-05"
  >
    <span
      className="text-left"
    >
      5
    </span>
  </button>
</div>
`;
